src/System.Linq/src/System/Linq/Reverse.cs

```diff
@@ -39,19 +38,25 @@ public override Iterator<TSource> Clone()
 
             public override bool MoveNext()
             {
+                if (_state - 2 <= -2)
```

Comment:
I think we need a short explanation of all the values of the state, so that this will be easier to maintain. I.E.
// 1 - iteration has just started (MoveNext wraps src in a buffer and sets _state to count +2
// [2 : int.MinValue +1]  - valid iteration state. src index is  _state - 3, so that we start with  "cnt + 2 - 3" -> cnt -1 and iterate downwards
// [int.minValue +2: 0]  - invalid/completed state. MoveNext calls Dispose and returns false. 

Similar comment should be added to the other modified MoveNexts